feat(efb): TOD calculator

arturkuma opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
Main purpose: implement TOD (top of descend) calculator. 

On the occasion:
- Add new components (button, card, divider and update input component)
- Implement react redux

## Screenshots (if necessary)

**New screenshots in [comment](https://github.com/flybywiresim/a32nx/pull/3425#issuecomment-779520523)**

Old ones:

![2021-02-08 20_46_00-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/107273601-903fe800-6a4f-11eb-8ef6-8562b1de8300.png)
![2021-02-08 20_46_43-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/107273609-933ad880-6a4f-11eb-8dd7-2b5a0acc043c.png)
![2021-02-08 20_47_22-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/107273640-9c2baa00-6a4f-11eb-89d0-9956c5c76222.png)
![2021-02-08 20_47_42-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/107273649-9e8e0400-6a4f-11eb-9e07-796c72ec44ac.png)


## References
<!-- You should be making changes based on some kind reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

Manual calculations that led to the creation of class `src/Service/TODCalculator.tsx`.

Alghoritm:
- get GS layers provided by the user
- calculate average GS based on time user will spend in each of that layers
- knowing average GS we know how much it will take to reach specific distance - so we can calculate vertical speed
- knowing average GS and vertical speed we know how much it will take to reach specific altitude - so we can calculate distance needed

## Additional context
<!-- Add any other context about the pull request here. -->

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub): arturkuma#6639

## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->
1. Check if *Tools* tool bar item is displayed
1. Make sure that calculations are correct
1. Ground speed works in ranges:

![2021-02-08 21_04_37-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/107274837-3e985d00-6a51-11eb-873a-bf84fe143b28.png)

in that configuration GS is:
- 250 KN between 0 and 10 000 FT
- 360 KN between 10 000 and 20 000 FT
- 436 KN above 20 000

order of GS layers is not important - alghoritm will sort it out

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

Benjozork commented:
Could this not go to the "Performance" tab ?

arturkuma commented:
> Could this not go to the "Performance" tab ?

There is some work in progress on "Performance" tab #3350 

I would leave it here in "Tools" for now until that other thing is merged to avoid conflicts

ExampleWasTaken commented:
The abbreviation M for nautical miles is used in marine context. ICAO standard is NM which is why I think we should use that too.

Source: https://en.m.wikipedia.org/wiki/Nautical_mile

donstim commented:
> The abbreviation M for nautical miles is used in marine context. ICAO standard is NM which is why I think we should use that too.
> 
> Source: https://en.m.wikipedia.org/wiki/Nautical_mile

It should also be "Start your descent" not "Start your descend."

donstim commented:
What exactly is to be input by the pilot here and what is either drawn from the flight plan or calculated?  I hope you are not expecting the pilot to be supplying the ground speed at each "layer."  I suggest simplifying this to either have it access https://descent.now.sh/social.png or be more similar to it.

ExampleWasTaken commented:
> 
> 
> What exactly is to be input by the pilot here and what is either drawn from the flight plan or calculated? I hope you are not expecting the pilot to be supplying the ground speed at each "layer." I suggest simplifying this to either have it access https://descent.now.sh/social.png or be more similar to it.

simBrief provides a calculated GS for every waypoint so maybe that could be used.

Benjozork commented on src/instruments/src/EFB/Components/Button/Button.tsx:
```diff
@@ -0,0 +1,51 @@
+    className?: any
+};
+
+function Button({ text, type = BUTTON_TYPE.BLUE, onClick, className, ...props }: props) {
```
Can we make this an `export const` ?

arturkuma replied:
done

Benjozork commented on src/instruments/src/EFB/Components/Card/Card.tsx:
```diff
@@ -0,0 +1,37 @@
+    title?: string
+};
+
+function Card({title, children, ...props}) {
```
Same here

arturkuma replied:
done

Benjozork commented on src/instruments/src/EFB/Components/Divider/Divider.tsx:
```diff
@@ -0,0 +1,28 @@
+
+import './Divider.scss'
+
+function Divider({className, ...props}) {
```
Same here

arturkuma replied:
done

Benjozork commented on src/instruments/src/EFB/Tools/Tools.tsx:
```diff
@@ -0,0 +1,42 @@
+    {name: 'TOD Calculator', renderComponent: () => <TODCalculator />}
+];
+
+function Tools() {
```
Same here

arturkuma replied:
done

Benjozork commented on src/instruments/src/EFB/TODCalculator/TODCalculator.tsx:
```diff
@@ -0,0 +1,36 @@
+import Data from "./Data/Data";
+import Result from "./Result/Result";
+
+function TODCalculator() {
```
Same here

arturkuma replied:
done

Benjozork commented on src/instruments/src/EFB/TODCalculator/Result/Result.tsx:
```diff
@@ -0,0 +1,70 @@
+import {TOD_CALCULATION_TYPE} from '../../Enum/TODCalculationType.enum';
+import {isNil} from 'lodash';
+
+function Result({currentAltitude, targetAltitude, calculation, groundSpeed, ...props}) {
```
Same here

arturkuma replied:
done

Benjozork commented on src/instruments/src/EFB/TODCalculator/GroundSpeed/GroundSpeed.tsx:
```diff
@@ -0,0 +1,73 @@
+import './GroundSpeed.scss'
+import Button, {BUTTON_TYPE} from "../../Components/Button/Button";
+
+function GroundSpeed({groundSpeed, addTodGroundSpeed, removeTodGroundSpeed, setTodGroundSpeed, ...props}) {
```
Same here

arturkuma replied:
done

Benjozork commented on src/instruments/src/EFB/TODCalculator/Data/Data.tsx:
```diff
@@ -0,0 +1,78 @@
+import {setTodData} from "../../Store/action-creator/tod-calculator";
+import {TOD_CALCULATION_TYPE} from "../../Enum/TODCalculationType.enum";
+
+function Data({
```
Same here

arturkuma replied:
done

arturkuma replied:
i think we can add this rule https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prefer-stateless-function.md in some of future changes

arturkuma commented:
> The abbreviation M for nautical miles is used in marine context. ICAO standard is NM which is why I think we should use that too.
> 
> Source: https://en.m.wikipedia.org/wiki/Nautical_mile

Right, fixed



> > The abbreviation M for nautical miles is used in marine context. ICAO standard is NM which is why I think we should use that too.
> > Source: https://en.m.wikipedia.org/wiki/Nautical_mile
> 
> It should also be "Start your descent" not "Start your descend."

Fixed :)

arturkuma commented:
> > What exactly is to be input by the pilot here and what is either drawn from the flight plan or calculated? I hope you are not expecting the pilot to be supplying the ground speed at each "layer." I suggest simplifying this to either have it access https://descent.now.sh/social.png or be more similar to it.
> 
> simBrief provides a calculated GS for every waypoint so maybe that could be used.

1. I'll think about simBrief integration after that. Let's keep it simple first. But thats a great idea and I'll create feature request once it's merged
1. 
- `What exactly is to be input by the pilot here and what is either drawn from the flight plan or calculated?` - pilot has to enter GS he thinks it'll be. He can take it from metar data, ATIS or wherever he wants. Usually pilots expect what speed they'll move with. Remember that this tool is not source of truth, it has to give some idea when to start descending.
-  `I hope you are not expecting the pilot to be supplying the ground speed at each "layer."` - As above. Pilot can add only one layer if he find it more intuitive and knows average speed. More layers = better accuracy. In your design it's similar (start speed, end speed)

Benjozork reviewed (approved):
Code LGTM

Benjozork commented:
@donstim can you take a look at this for the calculations side ?

arturkuma commented:
Note for me for later:
Create feature request - keep in sync current altitude and ground speed with actual ones (maybe option in settings as it's not realistic?)

donstim commented:
> @donstim can you take a look at this for the calculations side ?

Well, the calculations are "okay" except that the descent rate units are incorrectly displayed as ft/sec when they are actually ft/min. I would hate to have to descend at +3000 ft/sec!

But I have a number of issues with the proposed implementation. One is that this is an exact calculation where if you provide exact ground speed data throughout the descent and then fly exactly the vertical speed given, then you should arrive at the desired target altitude at the right point. But there is no margin for any error in any of that.

Personally, I'm just not a fan of the UI. Who is going to know their ground speed throughout the descent? You would normally fly a constant Mach, followed by constant KCAS so your true airspeed will be increasing in the constant Mach portion and decreasing in the constant KCAS portion. The ground speed will be the true airspeed minus any headwind (or plus any tailwind). Who is going to have any idea what their true airspeed is going to be during the descent let alone what the winds are going to be?

If it was my choice, I would have it request the speed schedule to be flown, for example, M0.78 to crossover altitude, then 280 KCAS to 10,000 feet, then 250 KCAS. Have the user input winds, but make sure they understand that wind changes will affect the estimate. Allow the speeds to either by input manually or pick them up from the MCDU to allow flying a managed CI speed descent. Then let the calculator do the drudge work of converting the speeds to KTAS and then to GS; something that calculators are good at. And provide a margin of something like 10 NM to allow for winds to be different than forecast and for additional slowing to an initial approach speed.

I'm also not a big fan of either requiring a vertical speed input to get the TOD point, or providing a vertical speed to reach the target point. Again, my preference here would be just to use a 3 degree descent gradient (in the background) and allow (or expect) the pilot to fly an open descent profile rather than a constant V/S descent.  To me, that would be more realistic until we have a true managed descent capability. My understanding is that a 3 degree descent prediction with accountability for winds and a 10 NM margin should work pretty well and allow for decelerating to an initial approach speed. (But if this was to be implemented, it would have to be checked.) Of course, a constant V/S descent could be given as an option.

My $0.02...

arturkuma commented:
> Well, the calculations are "okay" except that the descent rate units are incorrectly displayed as ft/sec when they are actually ft/min. I would hate to have to descend at +3000 ft/sec!

Whoopsie my bad, fixed

arturkuma commented:
Hi,

all what you've said @donstim makes sense of course and is great another approach, but I'm not sure about few things:

> One is that this is an exact calculation where if you provide exact ground speed data throughout the descent and then fly exactly the vertical speed given, then you should arrive at the desired target altitude at the right point. But there is no margin for any error in any of that.

Please remember that this is not ND display. It has to help pilot to decide, not decide for him. Any pilot will understand that and know that many factors may change during descend. At this point maybe we should add word 'about' before calculation output, like 'Start your descend about X NM before target'

>  I would have it request the speed schedule to be flown, for example, M0.78 to crossover altitude, then 280 KCAS to 10,000 feet, then 250 KCAS. Have the user input winds, but make sure they understand that wind changes will affect the estimate.

Agree there, it's easier to predict your KCAS than GS. On the other hand amount of things to input will be much bigger - KCAS, wind speed, wind direction. I think it's a good option but maybe it should be switchable - if user wants to enter GS or KCAS with all wind data

OstC6 commented:
Hello, if I may, the correct knot designation (the speed expressed in nautical mile) is "KT" not "KN".

arturkuma commented:
> Hello, if I may, the correct knot (the speed expressed in nautical mile) designation is "KT" not "KN".
> 
> Nice work aniway ;-)

fixed, thanks buddy

beheh commented on src/instruments/src/EFB/TODCalculator/Data/Data.tsx:
```diff
@@ -0,0 +1,78 @@
+}) => {
+    const calculationTypes = [
+        {label: 'Distance', rightLabel: 'NM', type: TOD_CALCULATION_TYPE.DISTANCE},
+        {label: 'Vertical speed', rightLabel: 'FT/m', type: TOD_CALCULATION_TYPE.VERTICAL_SPEED}
```
![image](https://user-images.githubusercontent.com/929769/107510063-bd051400-6ba3-11eb-911c-cf91c9d0c10b.png)

I would suggest following the ICAO definition and use exactly `ft/min` (in lowercase).

The uppercase `NM` seems correct:

![image](https://user-images.githubusercontent.com/929769/107510190-e0c85a00-6ba3-11eb-83ef-243cf4005c83.png)



Source: https://www.theairlinepilots.com/forumarchive/quickref/icao/annex5.pdf

arturkuma replied:
good point! changing

beheh commented on src/instruments/src/EFB/TODCalculator/Data/Data.tsx:
```diff
@@ -0,0 +1,78 @@
+                label={'Current altitude'}
+                type={'number'}
+                className={'dark-option mb-4'}
+                rightComponent={<span className={'text-lg'}>FT</span>}
```
Similar here and for the other inputs I would suggest we just stick with the lowercase `ft` as per the ICAO definition.

arturkuma replied:
fixed

beheh commented on src/instruments/src/EFB/TODCalculator/GroundSpeed/GroundSpeed.tsx:
```diff
@@ -0,0 +1,73 @@
+                            label={'GS'}
+                            type={'number'}
+                            className={'dark-option w-6/12'}
+                            rightComponent={<span className={'text-lg'}>KT</span>}
```
ICAO prefers the lowercase `kt`:

![image](https://user-images.githubusercontent.com/929769/107510517-58968480-6ba4-11eb-92c7-d1da790b6336.png)

(https://www.theairlinepilots.com/forumarchive/quickref/icao/annex5.pdf)

arturkuma replied:
fixed

beheh commented on src/instruments/src/EFB/TODCalculator/GroundSpeed/GroundSpeed.tsx:
```diff
@@ -0,0 +1,73 @@
+                            type={'number'}
+                            className={'dark-option w-6/12 mr-4'}
+                            value={from}
+                            rightComponent={<span className={'text-lg'}>FT</span>}
```
Similarly here, a lowercase `ft` would match ICAO.

arturkuma replied:
fixed

beheh commented on src/instruments/src/EFB/TODCalculator/Data/Data.tsx:
```diff
@@ -0,0 +1,78 @@
+                label={'Target altitude'}
+                type={'number'}
+                className={'dark-option mb-6'}
+                rightComponent={<span className={'text-lg'}>FT</span>}
```
And here, a lowercase ft would match ICAO.

arturkuma replied:
fixed

MisterChocker commented:
To add to @donstim

Normally you decent in open or managed mode. (V/S is normally used to maintain a given v/s to meet altitude restrictions or before reaching target altitudes.

The formula I learned doesn't need to know your exact GS, however you need to to the calculation every couple of thousand feet to check for errors.

The very basic version is altitude / 1000 * 3. (or in atc terms, 3nm for every 1000ft)
In addition to that you can then add the speed factor (IAS > 200: 1nm for every 10kt)
In addition to that you can then add the wind factor.

If this is done correctly the (I only know the boing term for this) banana (fms predicted level off indication) will be close to our calculations.

The above proposed equation is very simple but effective.

marcelago commented:
Thats an amazing quality-of-life inprovement, thanks for your contribution. I have a small suggestion: Have you considered making a similar calculator, that determines the correct V1,Vr and V2 speeds? I found that these types of calculators are hard to find and quite complex, but the efb could take most variables from the metadata of the plane itself- making things a lot simpler. (sorry, I know this is not the right thread- It just came to my mind, reading this thread :))

ZigTag commented:
React seems totally overkill for this

Benjozork commented:
@ZigTag it introduces Redux in the project so we can use it more later

ZigTag commented:
I think we should split up the PRs

ZigTag commented:
> I think we should split up the PRs

Disregard

ZigTag commented on src/instruments/src/EFB/Service/TODCalculator.tsx:
```diff
@@ -0,0 +1,107 @@
+type groundSpeed = {from: number, groundSpeed: number};
+type groundSpeedRange= {groundSpeed: number, range: [number, number]};
+
+export default class TODCalculator {
```
Do you think you can make this a function component?

arturkuma replied:
It's not a component, it's just a calculation class. yes it can be splitted in functions, but can't see any pros of that

ZigTag replied:
Ah apologies, I haven't seen a react class in a while and must have mistaken it for one. It's fine as is

ZigTag commented on src/instruments/src/EFB/TODCalculator/GroundSpeed/GroundSpeed.tsx:
```diff
@@ -0,0 +1,73 @@
+import './GroundSpeed.scss'
+import Button, {BUTTON_TYPE} from "../../Components/Button/Button";
+
+const GroundSpeed = ({groundSpeed, addTodGroundSpeed, removeTodGroundSpeed, setTodGroundSpeed, ...props}) => {
```
Make the props a type

arturkuma replied:
there is no props passed to this component from _outside_. Props are passed there from redux, so everything is within same file. props typing won't bring there any advantage I think

ZigTag replied:
Oh I thought it'd just clean up code, seeing that it is a redux thing, would it even be possible?

ZigTag commented on src/instruments/src/EFB/TODCalculator/Result/Result.tsx:
```diff
@@ -0,0 +1,74 @@
+import Card from "../../Components/Card/Card";
+import {TOD_CALCULATION_TYPE} from '../../Enum/TODCalculationType.enum';
+
+const Result = ({currentAltitude, targetAltitude, calculation, groundSpeed, ...props}) => {
```
Here as well

arturkuma replied:
as above

ZigTag commented:
Is the PR ready for testing @arturkuma ?

arturkuma commented:
Yes, just rebased in on master

St54Kevin commented:
It should read **NM** not M. Those are two different distance measurements

Benjozork commented:
@ZigTag see discussion on discord...

arturkuma commented:
@St54Kevin 

> It should read **"NM"** not "M". Those are two different distance measurements. Also "M" is used for metric meter, miles should be "Mi".

that was fixed few days ago

ZigTag commented:
I think you should update the screenshots when you get a chance

arturkuma commented:
@ZigTag sure, I'll do that in a min :)

arturkuma commented:
Updated screnshots:

![2021-02-15 10_25_13-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/107928218-58b5cc00-6f78-11eb-815f-eaca50f430d6.png)
![2021-02-15 10_25_37-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/107928229-5bb0bc80-6f78-11eb-9bfc-4a3247db2ef2.png)
![2021-02-15 10_25_54-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/107928239-5e131680-6f78-11eb-9236-a03ed14e9923.png)
![2021-02-15 10_26_22-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/107928253-60757080-6f78-11eb-9b62-66d2eede9883.png)

marcman86 commented:
Name: marcman86#4907
Date of testing: 15.2.2021
Version of the sim: 1.12.13.0
PR Tested: #3425
Tier of testing: 1
Changes to observe: PFD TOD tool
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP
2.) Turned on bat., ext. power and EFB
3.) Put in data and test the normal behaviour (Image1)
4.) Test some edge cases (Image2-4)



Media:

![64](https://user-images.githubusercontent.com/78707844/107939347-e4365980-6f86-11eb-8c13-5acf944d5a0c.png)
Image1

![65](https://user-images.githubusercontent.com/78707844/107939356-e7314a00-6f86-11eb-8305-adcddeb91b60.png)
Image2

![66](https://user-images.githubusercontent.com/78707844/107939370-ea2c3a80-6f86-11eb-9637-72aa1bef9e0f.png)
Image3

![67](https://user-images.githubusercontent.com/78707844/107939381-ec8e9480-6f86-11eb-9679-79351b412295.png)
Image4

Core mechanic works.

Layout:
- Can you make the 3 coloums the same widht (and extend the lighter grey background to (almost) the bottom)? Looks to patchy to me. (Like what you would have if you have 7 layers of ground speed)

Issues: 
- A few restrictions in the input would help (GS -250kt, Distance -80 NM, so on)
- TOD Calculator, but it is possible to put in a current altitude that is smaller then the target altitude
- You can scroll with the mousewheel zu change the inputs, but only in one direction. The other direction will zoom out.
- Scrolling large numbers like altitude is in single digit, should be in 100ft
- Adding a new GS layer will give no result while erasing a layer will set the result to 0 

Overall Rating: -

Conclusions of the testing: Needs to be improved.

Hope this helps :-)

St54Kevin commented:
Im unable to use it at all. I can input current and target ALT, but nothing happens. Also I cannot input a value into the left ALT field.

![image](https://user-images.githubusercontent.com/448627/107940871-34162000-6f89-11eb-8437-412e94b2cf35.png)

arturkuma commented:
@marcman86 thank you, will take a look at that.

Regarding

> Can you make the 3 coloums the same widht (and extend the lighter grey background to (almost) the bottom)? Looks to patchy to me. (Like what you would have if you have 7 layers of ground speed)

there is a reason behind not same-size columns. I wanted to make result as big as possible, while inputs have as much space as they need. By making same width columns we'll decrease readability of result and give extra space to inputs, but they dont need it.

About lighter gray - I'm afraid that this will be more patchy. But I'll try to do something with it

arturkuma commented:
@St54Kevin 

> Im unable to use it at all. I can input current and target ALT, but nothing happens. Also I cannot input a value into the left ALT field.

Hi you need to fill all required data to have some result.
In your screenshot you're missing distance OR vertical speed. You're also missing ground speeds. But remember that you can't change first altitude
> 
> ![image](https://user-images.githubusercontent.com/448627/107940871-34162000-6f89-11eb-8437-412e94b2cf35.png)

St54Kevin commented:
> @St54Kevin
> 
> > Im unable to use it at all. I can input current and target ALT, but nothing happens. Also I cannot input a value into the left ALT field.
> 
> Hi you need to fill all required data to have some result.
> In your screenshot you're missing distance OR vertical speed. You're also missing ground speeds. But remember that you can't change first altitude

You dont need these to calculate a TOD. Why do I need the distance and VS? Also i dont need ground speed if I select it manually anyways afterward. I thought this calculator is supposed to give me the groundspeed? 
And if so, I wasnt able to put in an ALT into the top left field anyways

marcman86 commented:
The result is big enough. the other fields are small. i mean just like the boxes are now but make them equal. Its just my opinion, maybe also wait for ohter reviews.

Something like this (Fastly drwan in Paint ;-P)...
![68](https://user-images.githubusercontent.com/78707844/107942590-bdc6ed00-6f8b-11eb-937d-e4d740632d9b.png)

arturkuma commented:
> > @St54Kevin
> > > Im unable to use it at all. I can input current and target ALT, but nothing happens. Also I cannot input a value into the left ALT field.
> > 
> > 
> > Hi you need to fill all required data to have some result.
> > In your screenshot you're missing distance OR vertical speed. You're also missing ground speeds. But remember that you can't change first altitude
> 
> You dont need these to calculate a TOD. Why do I need the distance and VS? Also i dont need ground speed if I select it manually anyways afterward. I thought this calculator is supposed to give me the VS?
> And if so, I wasnt able to put in an ALT into the top left field anyways.
> 
> I personally just calculate my TOD with both ALTs. And then descent with a 3 degree profile (VS=GS*10/2).
> 
> You are making it way more complicated than it should be

As I said earlier - you need _distance_ **OR** _vertical_ speed, not both :)

And of course you need your speed (no matter if it's GS or TAS - but GS is more precise here). Your TOD will be different if you're moving with 120kt vs 290kt. Later we're improve that so speed will be taken from airplane data, so pilot won't have to put it manually.

> I wasnt able to put in an ALT into the top left field anyways

like I said, first alt can't be changed.

On hight level altitudes you won't use descend degree angle too often. You'll use managed descend or vertical speed. The reason is X degree descend is ok on final where altitude, speed and wind is low (usually). Forcing aircraft to descend with 3 degree angle on hight level altitudes is not reliable. You'll reach desired point too soon/too late because of wind/aircraft weight/static pressure etc

marcman86 commented:
You can implement the general rule when no distance or vertical speed is entered and change it to your formula when its entered.

((Alt Diff[in ft]*3)/1000)+10[NM] it was, i think

So 30000ft gives 100NM

St54Kevin commented:
Honestly people will not know how to use your system at all I think. Even I dont get it. I can calulate the TOD in my head in about 5 secs. It needs to be way more simple as stated by @marcman86 above.

St54Kevin commented:
Or number the input fields so we know what to put in first. For example:

1. Current ALT
2. Dest ALT
 
System will then display a simple ToD distance with the info "manual VS management"

3. current GS
4. Dest GS

System will additonally display the correct VS

In my opinion we DONT need an input field for VS or distance because this is a TOD calculator not a descent manager, right? If I need to input a distance or my VS, Im past my TOD anyways

marcman86 commented:
I think what arturkuma does is (Alt diff[in ft] / VS[in ft/min])*Speed[in NM/min] and so on.

arturkuma commented:
Allright, I'll add basic version and keep current as more advanced one.

Basic version will ask user for current and target altitude. Optionally user will provide distance OR vertical seed like now.

> In my opinion we DONT need an input field for VS or distance because this is a TOD calculator not a descent manager, right?

I must disagree here. It can be optional of course, but I find it useful many times when I want to descend to some point with f.ex. -1500ft/m VS. 

Or when I have to reach some point, but doing it by open descend will lower my nose too much and I'll level my flight 20M before waypoint.

marcman86 commented:
Maybe in this way...
![69](https://user-images.githubusercontent.com/78707844/107949889-3763d880-6f96-11eb-8f76-d41e52969c2d.png)

kberg93 commented:
Category: QA Tester
Name: Kberg#0666
Date of testing: Current date 02/15/2021
Version of the sim: 1.12.13.0
PR Tested: #3425 
Tier of testing: 1
Changes to observe:
TOD Calculator added in efb
Testing technique: 
Using it to calculate TOD
Steps you've made to test the Pull Request: 
1. C&D LFMN set up plane and do the flight.
2. Put in data in TOD Calculator.
3. Start the decent.
4. Landing in DTTA.

Flight notes (When done a full flight):
No issues with the calculator, the Calculated V/S was pretty much on point as i can tell.

Media: Insert screenshots, recordings, or anything relevant to the review here.
![FlightSimulator_2rRDkOiuMD](https://user-images.githubusercontent.com/56963782/107981273-06999880-6fc2-11eb-97e1-26d079fa1266.png)

Issues (If Any): Only a small issue with A/P turning left or right not following the flightplan or heading now and then on the decent but thats nothing that should be caused by this PR as far as i know.

Overall Rating: Good.
Conclusions of the testing: From what i can tell this tool is good and works well.

arturkuma commented:
Guys!

I did some changes according to your comments, list below:
1. Current altitude can be in sync with SIM right now
1. Current ground speed as well
1. I've added 'Descend flight path angle' functionality - by giving **CURRENT ALTITUDE**, **TARGET ALTITUDE** and **DESCEND DEGREE** calculator will output **NM**. That was requested few times
1. By giving **remaining distance**, **current** and **target** altitude calculator will now give us **VS** like before and **desired angle**

I also did some layout changes according to @marcman86 comments. Screenshots below:

![2021-02-16 01_36_39-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/108006176-0e296380-6ffb-11eb-8e71-7b86028ad2a5.png)
![2021-02-16 01_36_54-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/108006180-108bbd80-6ffb-11eb-99c6-3cc1a9b0acee.png)
![2021-02-16 01_37_13-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/108006181-11bcea80-6ffb-11eb-927a-6db0f5d465f7.png)
![2021-02-16 01_37_35-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/108006183-12ee1780-6ffb-11eb-9c2b-611de956dba3.png)
![2021-02-16 01_37_56-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/108006184-1386ae00-6ffb-11eb-8feb-d3263775eb62.png)


I've also tested version with full height columns, but personally - I don't like it anyway.

![2021-02-16 01_50_59-Microsoft Flight Simulator - 1 12 13 0](https://user-images.githubusercontent.com/13150438/108006244-3b761180-6ffb-11eb-9340-57d658cae035.png)

So it's not in this bundle yet.

According some issues with inputs and restrictions on them, f.ex:

> You can scroll with the mousewheel zu change the inputs, but only in one direction. The other direction will zoom out.
Scrolling large numbers like altitude is in single digit, should be in 100ft

TOD calculator uses input component, which is not a subject of this PR. And this component can be improved.

We have a plan to implement virtual keyboard, so all issues related to that will be solved later. I'll work on that after TOD is approved and merged.

Kacper848 commented:
This looks great, nice work! Personally, I like this layout more:
![108006176-0e296380-6ffb-11eb-8e71-7b86028ad2a5](https://user-images.githubusercontent.com/65775520/108007849-d6b8b800-6ff6-11eb-88d3-c40dba6b6d8b.png)
The full height columns don't really work well IMO.

marcman86 commented:
Very cool! Will test it tomorrow.

From first view:

Did you include some savety buffer to the glideslope distance or substract it to calculate the GS angle?
Like 320 Sim Pilot mentioned:
https://www.youtube.com/watch?v=7ElFBOUhnMQ
at 13:20min

Also from just your images its not clear how to change the one field fron ° to ft/min to NM.
Is there a clickzone on the unit? If yes, can you mark it?

St54Kevin commented:
Well done! See, now we got something way more usable and better looking. Design wise I would make all three clumn the same length and vertically center the left and right one (when in auto mode).

@marcman86 That correction is only done for an approach intercept. This TOD calculator is not for an approach. And its also not good for descents less than 5000ft. Its too unprecise for such things. I personally add 5nm to all my calculations to have a little buffer.

arturkuma commented:
Another update :)

> Design wise I would make all three clumn the same length and vertically center the left and right one (when in auto mode).

I did it like proposed

> Also from just your images its not clear how to change the one field fron ° to ft/min to NM.
Is there a clickzone on the unit? If yes, can you mark it?

You just have to clear input, the other inputs will pop out. I've added button for that so it's more intuitive.

> Did you include some savety buffer to the glideslope distance or substract it to calculate the GS angle?

No, calculation is exact. But result states 'Start your descend about X NM before target'. Experienced pilot should know that hopefully :D

marcman86 commented:
Very nice!!! can you make the red button also blue? Red with a "x" looks like clear input.

arturkuma commented:
> Very nice!!! can you make the red button also blue? Red with a "x" looks like clear input. 
> 
> 
> 
> Maybe replace the x with a "v" (small arrow down).

Yes but this is how it works - it clears input

marcman86 commented:
> Yes but this is how it works - it clears input

Ah ok. But what if you start and dont have a input but want to switch to another unit? A new user wouldnt guess that.

arturkuma commented:
> > Yes but this is how it works - it clears input
> 
> 
> 
> Ah ok. But what if you start and dont have a input but want to switch to another unit? A new user wouldnt guess that.

When you start typing other inputs disapear and X button apears. I think thats a strong sign "hey, if you want to go to previous state just click the button that has just appeared"

arturkuma commented:
I think it's ready for another tier of QA? @ZigTag @Benjozork ?

marcman86 commented:
Name: marcman86#4907
Date of testing: 16.2.2021
Version of the sim: 1.12.13.0
PR Tested: #3425
Tier of testing: 1
Changes to observe: PFD TOD tool
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP
2.) Turned on bat., ext. power and EFB
3.) Put in alt data (Image1)
4.) Get to manual speed (Image2)
5.) Set V/S (Image3)

Media:

![72](https://user-images.githubusercontent.com/78707844/108054144-92f19d00-704e-11eb-9ccf-0cb815471df0.png)
Image1

![73](https://user-images.githubusercontent.com/78707844/108054171-9edd5f00-704e-11eb-9b9b-3382f14172dd.png)
Image2

![74](https://user-images.githubusercontent.com/78707844/108054181-a270e600-704e-11eb-881c-63f9a7a42079.png)
Image3

Issues: 
- No output on 3.) and 4.)
- Unexpected output in 5.)

Overall Rating: Needs to be improved

Conclusions of the testing:
- When entering current and target altitude the calculation should give a output ((Alt diff * 3) / 1000) when GS = 0kt (should display -kt in my opinion). 
- When switching to manual GS use 1 layer first. Then you got a usable output.
- On GS layers the naming Alt is not intuitive (maybe): Min. Alt 1, Min. Alt 2

Hope this helps :-)

arturkuma commented:
Thx for test :)

> Issues: No output on 3.) and 4.)

Thats correct behavior IMO, because you didnt put data needed to calculate - **DISTANCE** or **VS** or **ANGLE**

> Unexpected output in 5.)

Agree

> When switching to manual GS use 1 layer first. Then you got a usable output

If current altitude is > 10000ft then calculator will give you 2 layers by default. 
So when you're flying at 30000ft, calculator treats it as:
- 0 - 10 000 ft = 250 kt
- 10 000 ft - infinite = current GS from simulator

If you're below 10 000 then only one layer:
- 0 - 10 000ft = current GS from simulator

> On GS layers the naming Alt is not intuitive (maybe): Min. Alt 1, Min. Alt 2
I'll name it 'Bottom alt'

> When entering current and target altitude the calculation should give a output ((Alt diff * 3) / 1000) when GS = 0kt (should display -kt in my opinion).

I don't want to make special cases and assumes for GS, so the best option is to treat it as invalid input i think. So no result then

arturkuma commented:
@marcman86 you can recheck :)

- changed alt text (like proposed - Min. alt 1, Min. alt 2 etc)
- ground speeds with speed 0 will be treated like empty, if they are in range between <current alt, target alt>
- when switching from auto GS fetch to manual inputs:
    - if **current alt is set** and **it's > 10 000** and **current GS is > 250** - then create 2 layers
    - otherwise create one layer and put GS as it's value

marcman86 commented:
Testing: Its more intuitive right now. But it should give a first estimate by just typing in the altitude (for novice players).

Maybe instead of starting with auto updated GC, start with manual and 250kt and a -3 degree G/S as preset. Then you get a first estimate for TOD by just typing in the altitudes.

Minor things:
- The Coulums wobble in width when the third coloum comes in
- Call "Auto update" also "SYNC" like on the current altitude
- Center Auto update button like the Manual input button

After this it would be ready to go QA Tier 2 for my part! :-D

arturkuma commented:
> The Coulums wobble in width when the third coloum comes in

Fixed

> Call "Auto update" also "SYNC" like on the current altitude

Done

> Center Auto update button like the Manual input button

Did it


> -3 degree G/S as preset

-3 degree is now set as preset :)

> Maybe instead of starting with auto updated GC, start with manual and 250kt 

I think it's much better to have real value from SIM, instead of random value. 
And btw - GS is not needed when operating on flight path angle (which is by default -3 right now).

Thx :)

marcman86 commented:
Will QA testing when its reviewed and labeled "Ready to Test" :-)

arturkuma commented:
> Will QA testing when its reviewed and labeled "Ready to Test" :-)

@Benjozork @ZigTag ? 🙏

Benjozork commented:
Will review when we've dealt with MSFS update.

arturkuma commented:
FYI I've just squashed about 10/15 commits into 4 and rebased all of them to master

donstim commented:
Much improved! I like how you've given the pilot different options. After they try it a few times, they can adapt whatever method works best for them.

donstim reviewed (approved):
Much improved! I like how you've given the pilot different options. After they try it a few times, they can adapt whatever method works best for them.  For example, I chose a prediction based on a 3 degree descent angle, but then flew an open managed speed descent. I got to by target altitude about 25NM too early. So next time, maybe I'll try a lower descent angle, or a rate of descent that is about what I expect my average rate of descent to be.

Benjozork reviewed (approved):
LGTM

arturkuma commented:
@marcman86 I'm sure you want to check it again :D

marcman86 commented:
Name: marcman86#4907
Date of testing: 18.2.2021
Version of the sim: 1.13.16.0
PR Tested: #3425
Tier of testing: 2
Changes to observe: PFD TOD Calculator
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned in Rain C&D at EDDP
2.) Turned on bat. and ext. power
3.) Turned on PFB
4.) Go to Tools and PFD Calculator
5.) Tested different modes. Made sure the calculations go with the input.

Media:

![81](https://user-images.githubusercontent.com/78707844/108381000-af850500-7207-11eb-9660-2e4cb1439526.png)

Issues: None

Overall Rating: Good!

Conclusions of the testing: Calculations made sense for me. Like it very much. Very nice PR. Thank you for your work!

Hope this helps :-)

arturkuma commented:
thank you, and thank you all participants :) We've made useful tool, hope you'll like it

kberg93 commented:
Category: QA Tester
Name: Kberg#0666
Date of testing: Current date 02/18/2021
Version of the sim: 1.13.16.0
PR Tested: #3425 
Tier of testing: 1
Changes to observe:
TOD Calculator redesign since first version etc.

Testing technique: Testing the calculator.

Steps you've made to test the Pull Request: 
1. Spawn in and turn on the pad.
2. Go to Tools and TOD.
3. Test it out and see if it works.

Flight notes (When done a full flight):

Media: Insert screenshots, recordings, or anything relevant to the review here.
![FlightSimulator_cSDQNmiPZX](https://user-images.githubusercontent.com/56963782/108409932-15cd5000-7227-11eb-9f5a-84db534149f0.jpg)

Issues (If Any): None

Overall Rating: Good
Conclusions of the testing: Nice and helpfull tool to calculate TOD.

NonstopFlying commented:
Category: QA Trial
Discord: NonstopFlying #6771
Date of testing: Current date 02/18/2021
Version of the sim: 1.13.16.0
PR Tested: #3425
Tier of testing: 2
Changes to observe:
TOD Calculator

Testing technique: 
Testing the EFB tool for the TOD calculator

Steps you've made to test the Pull Request:

Spawned at EDDH
Turned on the EFB
Once at Curising I inputed the TOD stats and tested it out various times

Conclusion: Good
Worked like it said and I had no issues

BritishAvgeek commented:
Category: QA Tester
Name: BritishAvgeek#1436
Date of testing: 20/02/2021
Version of the sim: 1.13.16.0
PR Tested: #3425 
Tier of testing: 2
Changes to observe: TOD Calculator implementation within the EFB

Testing technique: Full Flight Quito to Lima livestreamed
Steps you've made to test the Pull Request:
1. Spawned C&D at Quito, set up using simbrief integration, checked general function of EFB and set that up with simbrief link too
2. Departed as normal out of Quito - had some issues NOT PR related
3. Planned for descent and made use of the EFB TOD for testing... automated input seemed bang on for multiple inputs
4. Tried Manual Input at FL370 in TOD calc - input using numpad almost triggered unrecoverable behaviour and the rudder and ailerons then went out of sync for about 5mins
5. Tested TOD for FL370 to FL100 initially, then FL370 to FL250 - then FL250 to FL100 then FL100 to 4000 and 4000 to 3000 - all worked very well on auto alt sync
6. Broke off on ILS15 for visual manual circling approach R33

Flight notes (When done a full flight):
SEQM - QIT3A KETOM UM674 ATATU ATAT2 set up for ILS15 then broke off for circling visual R33 approach due to wind.

Media: 
Livestreamed here https://youtu.be/dkWkPHGwPEU  (head to about 1hr 30mins onwards for using TOD calculator with issues etc)
![image](https://user-images.githubusercontent.com/71381591/108601164-200c5d00-7393-11eb-9407-14ffdeba9cd8.png)


Issues (If Any):
1. Manual Input frustrated by the keybind issue... it triggered hard rudder input inadvertently (my own fault I guess, I thought I'd cleared my keybinds) but even with data input it didn't update for me so not quite sure how to "work" that page. The keybind rudder attack was resolved by immediately deactivating autopilot, resetting rudder trim and then reactivating AP when the aircraft settled.
2. After settling, the aircraft kept a bank angle of about 2.5 to 5 degrees - cleared by selecting Selected Heading Mode, changing heading a few degrees then switching back to Managed Mode
3. One issue I did have was that the altitude kept DE-SYNCING itself. Notably at FL130 it got itself stuck there, so I had to reclick "Sync" to wake it up and get it back to FL100 which was my altitude at the time I re-synced. This de-syncing happened twice overall but only minor - the sync button sorted it.

No other issues - issues were really due to the keybind issue more than the PR though I couldn't seem to get the manual data input mode to change the calculated values.

Overall Rating: To be improved - I get the keybind issue isn't a problem with this PR as such so that is ignored in the "Overall Rating" sense... however, I'd look at the desyncing issues - not sure if that can be fixed? 

Conclusions of the testing: Great addition to the EFB - works really well, calculations helped and it was in a really easy to use format. Ground Speed remained synced the whole time - and altitude remained synced 90% of the time too. Great work, thanks!

arturkuma commented:
Altitude desync happens only when changing EFB tabs. TOD component looses its state.

However it's already fixed in another branch and it waits for this PR to be merged first :) so should be fixed shortly
